ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java

```diff
@@ -121,36 +133,38 @@ public Transaction createPrivacyMarkerTransaction(
   public ValidationResult<TransactionValidator.TransactionInvalidReason> validatePrivateTransaction(
       final PrivateTransaction privateTransaction, final String privacyGroupId) {
     return privateTransactionValidator.validate(
-        privateTransaction, getSenderNonce(privateTransaction.getSender(), privacyGroupId));
+        privateTransaction, determineNonce(privateTransaction.getSender(), privacyGroupId));
   }
 
-  private SendRequest createSendRequest(final PrivateTransaction privateTransaction) {
-    final BytesValueRLPOutput bvrlp = new BytesValueRLPOutput();
-    privateTransaction.writeTo(bvrlp);
-    final String payload = BytesValues.asBase64String(bvrlp.encoded());
+  public long determineNonce(
+      final String privateFrom, final String[] privateFor, final Address address) {
+    final String[] groupMembers = Arrays.append(privateFor, privateFrom);
 
-    if (privateTransaction.getPrivacyGroupId().isPresent()) {
-      return new SendRequestBesu(
-          payload,
-          enclavePublicKey,
-          BytesValues.asBase64String(privateTransaction.getPrivacyGroupId().get()));
-    } else {
-      final List<String> privateFor =
-          privateTransaction.getPrivateFor().get().stream()
-              .map(BytesValues::asBase64String)
-              .collect(Collectors.toList());
+    final FindPrivacyGroupRequest request = new FindPrivacyGroupRequest(groupMembers);
+    final List<PrivacyGroup> matchingGroups = Lists.newArrayList(enclave.findPrivacyGroup(request));
 
-      // FIXME: orion should accept empty privateFor
-      if (privateFor.isEmpty()) {
-        privateFor.add(BytesValues.asBase64String(privateTransaction.getPrivateFrom()));
-      }
+    final List<PrivacyGroup> legacyGroups =
+        matchingGroups.stream()
+            .filter(group -> group.getType() == Type.LEGACY)
+            .collect(Collectors.toList());
 
-      return new SendRequestLegacy(
-          payload, BytesValues.asBase64String(privateTransaction.getPrivateFrom()), privateFor);
+    if (legacyGroups.size() == 0) {
+      // the legacy group does not exist yet
+      return 0;
     }
+
+    if (legacyGroups.size() != 1) {
```

Comment:
nit; you can replace this if block by using `com.google.common.base.Preconditions` `checkArgument`, fewer lines of code, maybe nicer/easier to read